src/Reflection/ReflectionFunctionAbstract.php

```diff
@@ -619,4 +619,17 @@ public function getReturnStatementsAst()
 
         return $visitor->getReturnNodes();
     }
+
+    private function findTypeFromAst($namespace, $locatedSource, $type)
+    {
+        $objectType = (new FindTypeFromAst())->__invoke(
+            $this->reflector->getContextFactory()->createForNamespace(
```

Comment:
Probably - the thing that makes me hesitate is that the `SourceLocator` classes then have to be aware of both the `Reflector` and the `ContextFactory` (which belongs to the `Reflector`).

The `SourceLocator` classes are currently responsible for instantiating the reflection classes.

I wonder if it is possible to make the `Reflector` classes responsible for creating the `Reflection[Class|Function]` instances, and the `SourceLocator` responsible only for getting `LocatedSource` /cc @asgrim